pkg/distributor/distributor.go

```diff
@@ -769,7 +769,11 @@ func (d *Distributor) prePushHaDedupeMiddleware(next PushFunc) PushFunc {
 	}
 }
 
-func (d *Distributor) prePushRelabelMiddleware(next PushFunc) PushFunc {
+// prePushRelabelAndSortMiddleware is responsible for applying relabel rules as
+// well as sorting labels. The sort step is critical because ingesters assume
+// that labels are sorted in the push request. Without this, ingesters will
+// write a corrupted block index.
```

Comment:
Comment about sort step being critical is similar to comment https://github.com/grafana/mimir/blob/1af7e139bb21b8300b238691b98358948fedcc0c/pkg/distributor/distributor.go#L825-L830. Moving sorting into its own middleware can make that very clear and explicit.

Reply:
Historical note: "by all labels" is redundant, since sharding always looks at all labels.
There used to be an option to shard only on `__name__`.

Reply:
I've moved the sort to a separate middleware and updated that comment :+1: